engine/classes/Elgg/EntityCapabilitiesService.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace Elgg;
+
+/**
+ * Entity capabilities service
+ *
+ * @internal
+ * @since 4.1
+ */
+class EntityCapabilitiesService {
+	
+	/**
+	 * Stores the entities with their capabilities
+	 *
+	 * @var array
+	 */
+	protected $entities = [];
+	
+	/**
+	 * Checks if an entity has a specific capability
+	 *
+	 * @param string $type       type of the entity
+	 * @param string $subtype    subtype of the entity
+	 * @param string $capability name of the capability to check
+	 * @param bool   $default    default value to return if it is not explicitly set
+	 *
+	 * @return bool
+	 */
+	public function hasCapability(string $type, string $subtype, string $capability, bool $default = false): bool {
+		return elgg_extract($capability, elgg_extract($subtype, $this->entities[$type], []), $default);
```

Comment:
`$this->entities[$type]` will cause a PHP warning if `$type` isn't set in the array